src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,734 @@
+}
+
+void JfrCPUTimeTraceQueue::clear() {
+  Atomic::release_store(&_head, (u4)0);
+}
+
+static int64_t compute_sampling_period(double rate) {
+  if (rate == 0) {
+    return 0;
+  }
+  return os::active_processor_count() * 1000000000.0 / rate;
+}
+
+class JfrCPUSamplerThread : public NonJavaThread {
+  friend class JfrCPUTimeThreadSampling;
+ private:
+  Semaphore _sample;
+  NonJavaThread* _sampler_thread;
+  double _rate;
+  bool _auto_adapt;
+  volatile int64_t _current_sampling_period_ns;
+  volatile bool _disenrolled;
+  volatile bool _stop_signals;
+  volatile int _active_signal_handlers;
+  volatile bool _is_async_processing_of_cpu_time_jfr_requests_triggered;
+
+  JfrCPUSamplerThread(double rate, bool auto_adapt);
+
+  void start_thread();
+
+  void enroll();
+  void disenroll();
+  void update_all_thread_timers();
+
+  void auto_adapt_period_if_needed();
+
+  void set_rate(double rate, bool auto_adapt);
+  int64_t get_sampling_period() const { return Atomic::load(&_current_sampling_period_ns); };
+
+  void sample_thread(JfrSampleRequest& request, void* ucontext, JavaThread* jt, JfrThreadLocal* tl);
+
+  // process the queues for all threads that are in native state (and requested to be processed)
+  void stackwalk_threads_in_native();
+  bool create_timer_for_thread(JavaThread* thread, timer_t &timerid);
+
+protected:
+  virtual void post_run();
+public:
+  virtual const char* name() const { return "JFR CPU Sampler Thread"; }
+  virtual const char* type_name() const { return "JfrCPUTimeSampler"; }
+  void run();
+  void on_javathread_create(JavaThread* thread);
+  void on_javathread_terminate(JavaThread* thread);
+
+  void handle_timer_signal(siginfo_t* info, void* context);
+  bool init_timers();
+  void stop_timer();
+
+  void trigger_async_processing_of_cpu_time_jfr_requests();
+};
+
+// we have two stop signals, to remove a data race
+// and this works even when the _instance is null
+static volatile bool _static_stop_signals = true;
+
+
+JfrCPUSamplerThread::JfrCPUSamplerThread(double rate, bool auto_adapt) :
+  _sample(),
+  _sampler_thread(nullptr),
+  _rate(rate),
+  _auto_adapt(auto_adapt),
+  _current_sampling_period_ns(compute_sampling_period(rate)),
+  _disenrolled(true),
+  _stop_signals(true),
+  _active_signal_handlers(0),
+  _is_async_processing_of_cpu_time_jfr_requests_triggered(false) {
+  assert(rate >= 0, "invariant");
+}
+
+void JfrCPUSamplerThread::trigger_async_processing_of_cpu_time_jfr_requests() {
+  Atomic::release_store(&_is_async_processing_of_cpu_time_jfr_requests_triggered, true);
+}
+
+void JfrCPUSamplerThread::on_javathread_create(JavaThread* thread) {
+  if (thread->is_hidden_from_external_view() || thread->is_JfrRecorder_thread()) {
+    return;
+  }
+  JfrThreadLocal* tl = thread->jfr_thread_local();
+  assert(tl != nullptr, "invariant");
+  tl->cpu_time_jfr_queue().resize_for_period(_current_sampling_period_ns / 1000000);
+  timer_t timerid;
+  if (create_timer_for_thread(thread, timerid)) {
```

Comment:
Timer creation failure is not an impossible situation, we should somehow let user know that not all threads are being profiled but without flooding in logs. One warning per profiling session may be a good compromise.
You can verify failure condition by setting low `ulimit -i`.

Reply:
I added a "Failed to create timer for a thread" warning